src/geointerface.jl

```diff
@@ -23,57 +35,282 @@ let pointtypes = (wkbPoint, wkbPoint25D, wkbPointM, wkbPointZM),
+            GeoInterface.MultiLineStringTrait()
         elseif gtype in polygontypes
-            :Polygon
+            GeoInterface.PolygonTrait()
         elseif gtype in multipolygontypes
-            :MultiPolygon
+            GeoInterface.MultiPolygonTrait()
         elseif gtype in collectiontypes
-            :GeometryCollection
+            GeoInterface.GeometryCollectionTrait()
+        elseif gtype == wkbCircularString
+            GeoInterface.CircularStringTrait()
+        elseif gtype == wkbCompoundCurve
+            GeoInterface.CompoundCurveTrait()
+        elseif gtype == wkbCurvePolygon
+            GeoInterface.CurvePolygonTrait()
+        elseif gtype == wkbMultiSurface
+            GeoInterface.MultiSurfaceTrait()
+        elseif gtype == wkbPolyhedralSurface
+            GeoInterface.PolyhedralSurfaceTrait()
+        elseif gtype == wkbTIN
+            GeoInterface.TINTrait()
+        elseif gtype == wkbTriangle
+            GeoInterface.TriangleTrait()
         else
-            @warn "unknown geometry type" gtype
-            :Unknown
+            @warn "unknown geometry type"
+            nothing
         end
     end
 
-    function GeoInterface.coordinates(g::AbstractGeometry)
-        gtype = getgeomtype(g)
-        ndim = getcoorddim(g)
-        return if gtype in pointtypes
-            if ndim == 2
-                Float64[getx(g, 0), gety(g, 0)]
-            elseif ndim == 3
-                Float64[getx(g, 0), gety(g, 0), getz(g, 0)]
-            else
-                error("getcoorddim($g) returned $ndim: expected 2 or 3")
-            end
-        elseif gtype in multipointtypes
-            Vector{Float64}[
-                GeoInterface.coordinates(getgeom(g, i - 1)) for i in 1:ngeom(g)
-            ]
-        elseif gtype in linetypes || gtype == wkbLinearRing
-            Vector{Float64}[
-                collect(getpoint(g, i - 1)[1:ndim]) for i in 1:ngeom(g)
-            ]
-        elseif gtype in multilinetypes || gtype in polygontypes
-            Vector{Vector{Float64}}[
-                GeoInterface.coordinates(getgeom(g, i - 1)) for i in 1:ngeom(g)
-            ]
-        elseif gtype in multipolygontypes
-            Vector{Vector{Vector{Float64}}}[
-                GeoInterface.coordinates(getgeom(g, i - 1)) for i in 1:ngeom(g)
-            ]
+    function GeoInterface.is3d(::GeometryTraits, geom::AbstractGeometry)
+        return getcoorddim(geom) >= 3
+    end
+    function GeoInterface.ismeasured(::GeometryTraits, geom::AbstractGeometry)
+        return getcoorddim(geom) >= 4
+    end
+
+    function GeoInterface.ncoord(::GeometryTraits, geom::AbstractGeometry)
+        return getcoorddim(geom)
+    end
+
+    function GeoInterface.getcoord(::GeometryTraits, geom::AbstractGeometry, i)
+        if i == 1
+            getx(geom, 0)
+        elseif i == 2
+            gety(geom, 0)
+        elseif i == 3  # M is an option here, but not properly supported by ArchGDAL yet
+            getm(geom, 0)
```

Comment:
Out of curiosity, when is it Z, versus M?

Reply:
Well I made a mistake here, I think we should call `getz` here. It could be `m`, but only when `is3d` is false and `ismeasured` true?

Reply:
Addressed by https://github.com/yeesian/ArchGDAL.jl/pull/290/commits/96f9731f18daec06b1d9c07a98903917aa8fa8ce